cli/src/commands/config.rs

```diff
@@ -296,9 +297,56 @@ pub(crate) fn cmd_config_set(
             path = config_path.display()
         )));
     }
+
+    // If the user is trying to change the author config, we should warn them that
+    // it won't affect the working copy author
+    if args.name == ConfigNamePathBuf::from_iter(vec!["user", "name"]) {
+        let config = command.settings().config();
+        let user_name: String = config.get("user.name")?;
+
+        // only warn if the author is actually changing
+        if user_name != args.value {
```

Comment:
`args.value` may be escaped in TOML syntax, whereas `user_name` is a parsed value.

Maybe we can instead reload the config file and compare name/email fields? That might be silly, but we can reuse the same function for `config edit`.

Reply:
> Maybe we can instead reload the config file and compare name/email fields?

That was my initial idea, but it it also felt silly. I would basically have to have 2 entire copies of the config, one before and the other after.

> `args.value` may be escaped in TOML syntax, whereas `user_name` is a parsed value.

WDYM by this? I thought both are just string values. I mean, I tested it locally and it works as intended, what am I missing here?

Reply:
Actually, scratch that, it does not work as intended

Reply:
The bug is now fixed, but I'm still using `args.value`

Reply:
Try something like `jj config set user.name '"Foo"'`. That's the original issue I found.

Reply:
`jj config set user.name '"Foo"'` works for me by setting the user to `Foo`. Isn't that expected?

Reply:
> `jj config set user.name '"Foo"'` works for me by setting the user to `Foo`. Isn't that expected?

`args.value = '"Foo"'` is parsed as `Foo`, not `"Foo"` as you said, which means `user_name != args.value` doesn't work in that case. `user_name` is parsed `Foo` whereas `args.value` is raw `"Foo"`.

Perhaps, `write_config_value_to_file()` can be changed to take `value: toml_edit::Value` instead of `value_str: &str`.

Reply:
>  `user_name != args.value` doesn't work in that case

Ah now I understand. I thought not having quotes / parsing the string _was_ the intended behaviour.

> Perhaps, `write_config_value_to_file()` can be changed to take `value: toml_edit::Value` instead of `value_str: &str`.

I'll try to investigate in the upcoming days, but I'm not super confident. A naive change to `::Value` resulted in tests failing because everything ends up being a TOML string, because there's no toml parsing. But maybe I come up with a better way.

Reply:
https://github.com/martinvonz/jj/pull/4244

Parsing is intended. We just need to use parsed value for comparison.

Reply:
Hmm I rebased past your PR and it's still the same:

```
❯ PATH="~/InCogNiTo124-contrib/jj/target/debug:${PATH}"; which jj
~/InCogNiTo124-contrib/jj/target/debug/jj
❯ jj git init
Initialized repo in "."
❯ jj config set --repo user.email "'Bar'"
Warning: This setting will only impact future commits.
The author of the working copy will stay "InCogNiTo124 <msmetko@msmetko.xyz>". 
To change the working copy author, use "jj describe --reset-author --no-edit"
❯ jj describe --reset-author --no-edit
Working copy now at: pontmtzu 64dc19f3 (empty) (no description set)
Parent commit      : zzzzzzzz 00000000 (empty) (no description set)
❯ jj log
@  pontmtzu Bar 2024-08-09 20:53:21 64dc19f3
│  (empty) (no description set)
◆  zzzzzzzz root() 00000000
```

Reply:
Use parsed `value` instead of `args.value`.

Reply:
> Use parsed value instead of args.value.

still does not work as intended:
```console
❯ jj config set --repo user.email '"Foo"'
Warning: This setting will only impact future commits.
The author of the working copy will stay "InCogNiTo124 <msmetko@msmetko.xyz>". 
To change the working copy author, use "jj describe --reset-author --no-edit"
❯ jj describe --reset-author --no-edit
Working copy now at: vvtrswtz 84b3b3dc (empty) (no description set)
Parent commit      : zzzzzzzz 00000000 (empty) (no description set)
❯ jj log
@  vvtrswtz Foo 2024-08-10 13:50:54 84b3b3dc
│  (empty) (no description set)
◆  zzzzzzzz root() 00000000
❯ cat .jj/repo/config.toml
[user]
email = "Foo"
```

Reply:
What do you mean by "as intended"? The test case would be something like
1. create working-copy commit as email: `Foo`
2. `jj config set --repo user.email '"Foo"'`
And 2 shouldn't suggest `--reset-author` because the author email is unchanged.

Oh, and you'll have to use `.as_str()` instead of `.to_string()`. The latter will encode `toml::Value` as a TOML expression.
https://docs.rs/toml_edit/latest/toml_edit/enum.Value.html#method.as_str

`.as_str()` is stricter than `config.get_string()`, but that's probably okay.

Reply:
The latest push should work:

<details>
<summary>Test case</summary>

```console
❯ # SETUP
❯ jj git init
Initialized repo in "."
❯ jj log
@  nwlvpyok msmetko@msmetko.xyz 2024-08-12 19:17:16 1d8a1de3
│  (empty) (no description set)
◆  zzzzzzzz root() 00000000

❯ # EXPECTED NO WARNING (same author)
❯ jj config set --repo user.email "msmetko@msmetko.xyz"

❯ # EXPECTED WARNING (new author)
❯ jj config set --repo user.email "Foo"
Warning: This setting will only impact future commits.
The author of the working copy will stay "InCogNiTo124 <msmetko@msmetko.xyz>". 
To change the working copy author, use "jj describe --reset-author --no-edit"

❯ # EXPECTED AUTHOR RESET
❯ jj describe --reset-author --no-edit
Working copy now at: nwlvpyok 8c648ab7 (empty) (no description set)
Parent commit      : zzzzzzzz 00000000 (empty) (no description set)
❯ jj log
@  nwlvpyok Foo 2024-08-12 19:18:14 8c648ab7
│  (empty) (no description set)
◆  zzzzzzzz root() 00000000

❯ # EXPECTED NO WARNING (same author)
❯ jj config set --repo user.email "'Foo'"
❯ jj config set --repo user.email '"Foo"'
```

</details>